File: apps/ios/0dteTrader/Features/Chart/UltimateSupportResistance/UsrSettings.swift
import Foundation

/// Single source of truth for every numeric Pine input bound. The settings
/// validator and SwiftUI controls both consume these ranges so a UI change
/// cannot admit a value that persistence later rejects (or vice versa).
enum UsrSettingsBounds {
    static let proximityPercent: ClosedRange<Double> = 1...50
    static let maxSupportLevels = 1...500
    static let maxResistanceLevels = 1...500
    static let maxRecentSignalsTotal = 5...100
    static let volumeLookback = 10...200
    static let minimumRelativeVolume: ClosedRange<Double> = 1...5
    static let minimumVolumeZScore: ClosedRange<Double> = 0...5
    static let maxSequenceLength = 2...50
    static let displacementBodyPercent: ClosedRange<Double> = 40...95
    static let displacementAtrMultiplier: ClosedRange<Double> = 0.2...3
    static let structureLookback = 2...20
    static let pivotLeftBars = 1...10
    static let pivotRightBars = 1...5
    static let gapAtrMultiplier: ClosedRange<Double> = 0.05...3
    static let breakBufferTicks = 1...20
    static let zoneMitigationPercent: ClosedRange<Double> = 0.5...1
    static let minimumTick: ClosedRange<Double> = 0.000_001...100
    static let poolClusterThreshold = 2...10
    static let poolAtrFactor: ClosedRange<Double> = 1...5
    static let maxSupportPools = 1...60
    static let maxResistancePools = 1...60
    static let fvgFillPercent: ClosedRange<Double> = 10...100
    static let fvgLookback = 3...50
    static let fvgBodyPercent: ClosedRange<Double> = 0.05...3
    static let fvgWickPercent: ClosedRange<Double> = 0...2
    static let maxVisibleFvgs = 1...15
    static let fvgMaxBarsActive = 10...500
    static let fvgMinGapAtr: ClosedRange<Double> = 0...1
    static let fvgMinBodyAtr: ClosedRange<Double> = 0...3
}

/// Pine input contract for Ultimate Support & Resistance. Flat/Codable by
/// design: persisted settings can add fields while old payloads fall back to
/// the canonical defaults through SettingsStore.
struct UsrSettings: Codable, Equatable, Sendable {
    var enabled = false
    var enableProximityFilter = true
    var proximityPercent = 5.0
    var maxSupportLevels = 125
    var maxResistanceLevels = 125
    var showLiquidityPools = true
    var showFvg = true
    var analysisTimeframe = "chart"
    var customTimeframe = "60"
    var showConfluence = false
    var enableSrFlip = true
    var showBounceSignals = false
    var showSweepSignals = false
    var signalRequireQualification = true
    var requireConfirmationCandleDirection = true
    var cancelOpposingSignal = true
    var maxRecentSignalsTotal = 20
    var volumeLookback = 30
    var minimumRelativeVolume = 1.2
    var minimumVolumeZScore = 0.5
    var sessionAwareVolume = true
    var maxSequenceLength = 12
    var displacementBodyPercent = 60.0
    var displacementAtrMultiplier = 0.6
    var structureLookback = 5
    var pivotLeftBars = 3
    var pivotRightBars = 1
    var orderBlockUseWicks = false
    var gapAtrMultiplier = 0.3
    var requirePriceVoidGaps = true
    var breakBufferTicks = 1
    var zoneMitigationPercent = 0.95
    var minimumTick = 0.01
    var showFlippedOrigins = true
    var showAllBrokenLevels = false
    var hidePooledLines = true
    var poolClusterThreshold = 3
    var poolAtrFactor = 2.5
    var maxSupportPools = 30
    var maxResistancePools = 30
    var showIfvg = true
    var showFvgCe = true
    var showFvgLabels = false
    var fvgFillMode = "ce"
    var fvgFillPercent = 50.0
    var fvgLookback = 10
    var fvgBodyPercent = 0.36
    var fvgWickPercent = 0.5
    var maxVisibleFvgs = 5
    var fvgMaxBarsActive = 200
    var fvgMinGapAtr = 0.05
    var fvgMinBodyAtr = 0.5
    var fvgBullishColor = "rgba(1, 199, 31, 0.15)"
    var fvgBearishColor = "rgba(216, 0, 0, 0.15)"
    var fvgCeColor = "rgba(255, 235, 59, 0.30)"
    var ifvgBullishColor = "rgba(255, 152, 0, 0.15)"
    var ifvgBearishColor = "rgba(156, 39, 176, 0.15)"

    private enum CodingKeys: String, CodingKey {
        case enabled, enableProximityFilter, proximityPercent
        case maxSupportLevels, maxResistanceLevels, showLiquidityPools, showFvg
        case analysisTimeframe, customTimeframe, showConfluence, enableSrFlip
        case showBounceSignals, showSweepSignals, signalRequireQualification
        case requireConfirmationCandleDirection, cancelOpposingSignal, maxRecentSignalsTotal
        case volumeLookback, minimumRelativeVolume, minimumVolumeZScore, sessionAwareVolume
        case maxSequenceLength, displacementBodyPercent, displacementAtrMultiplier
        case structureLookback, pivotLeftBars, pivotRightBars, orderBlockUseWicks
        case gapAtrMultiplier, requirePriceVoidGaps, breakBufferTicks, zoneMitigationPercent
        case minimumTick, showFlippedOrigins, showAllBrokenLevels, hidePooledLines
        case poolClusterThreshold, poolAtrFactor, maxSupportPools, maxResistancePools
        case showIfvg, showFvgCe, showFvgLabels, fvgFillMode, fvgFillPercent
        case fvgLookback, fvgBodyPercent, fvgWickPercent, maxVisibleFvgs, fvgMaxBarsActive
        case fvgMinGapAtr, fvgMinBodyAtr, fvgBullishColor, fvgBearishColor, fvgCeColor
        case ifvgBullishColor, ifvgBearishColor
    }

    init() {}

    /// Decode additively so a settings payload written by an older app keeps
    /// every user choice it knows about while newly introduced fields receive
    /// their canonical defaults. A wrong-typed present value still fails
    /// closed and is rejected by SettingsStore.
    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        self.init()
        enabled = try values.decodeIfPresent(Bool.self, forKey: .enabled) ?? enabled
        enableProximityFilter = try values.decodeIfPresent(Bool.self, forKey: .enableProximityFilter)
            ?? enableProximityFilter
        proximityPercent = try values.decodeIfPresent(Double.self, forKey: .proximityPercent) ?? proximityPercent
        maxSupportLevels = try values.decodeIfPresent(Int.self, forKey: .maxSupportLevels) ?? maxSupportLevels
        maxResistanceLevels = try values.decodeIfPresent(Int.self, forKey: .maxResistanceLevels) ?? maxResistanceLevels
        showLiquidityPools = try values.decodeIfPresent(Bool.self, forKey: .showLiquidityPools) ?? showLiquidityPools
        showFvg = try values.decodeIfPresent(Bool.self, forKey: .showFvg) ?? showFvg
        analysisTimeframe = try values.decodeIfPresent(String.self, forKey: .analysisTimeframe) ?? analysisTimeframe
        customTimeframe = try values.decodeIfPresent(String.self, forKey: .customTimeframe) ?? customTimeframe
        showConfluence = try values.decodeIfPresent(Bool.self, forKey: .showConfluence) ?? showConfluence
        enableSrFlip = try values.decodeIfPresent(Bool.self, forKey: .enableSrFlip) ?? enableSrFlip
        showBounceSignals = try values.decodeIfPresent(Bool.self, forKey: .showBounceSignals) ?? showBounceSignals
        showSweepSignals = try values.decodeIfPresent(Bool.self, forKey: .showSweepSignals) ?? showSweepSignals
        signalRequireQualification = try values.decodeIfPresent(Bool.self, forKey: .signalRequireQualification)
            ?? signalRequireQualification
        requireConfirmationCandleDirection = try values.decodeIfPresent(
            Bool.self,
            forKey: .requireConfirmationCandleDirection
        ) ?? requireConfirmationCandleDirection
        cancelOpposingSignal = try values.decodeIfPresent(Bool.self, forKey: .cancelOpposingSignal)
            ?? cancelOpposingSignal
        maxRecentSignalsTotal = try values.decodeIfPresent(Int.self, forKey: .maxRecentSignalsTotal)
            ?? maxRecentSignalsTotal
        volumeLookback = try values.decodeIfPresent(Int.self, forKey: .volumeLookback) ?? volumeLookback
        minimumRelativeVolume = try values.decodeIfPresent(Double.self, forKey: .minimumRelativeVolume)
            ?? minimumRelativeVolume
        minimumVolumeZScore = try values.decodeIfPresent(Double.self, forKey: .minimumVolumeZScore)
            ?? minimumVolumeZScore
        sessionAwareVolume = try values.decodeIfPresent(Bool.self, forKey: .sessionAwareVolume) ?? sessionAwareVolume
        maxSequenceLength = try values.decodeIfPresent(Int.self, forKey: .maxSequenceLength) ?? maxSequenceLength
        displacementBodyPercent = try values.decodeIfPresent(Double.self, forKey: .displacementBodyPercent)
            ?? displacementBodyPercent
        displacementAtrMultiplier = try values.decodeIfPresent(Double.self, forKey: .displacementAtrMultiplier)
            ?? displacementAtrMultiplier
        structureLookback = try values.decodeIfPresent(Int.self, forKey: .structureLookback) ?? structureLookback
        pivotLeftBars = try values.decodeIfPresent(Int.self, forKey: .pivotLeftBars) ?? pivotLeftBars
        pivotRightBars = try values.decodeIfPresent(Int.self, forKey: .pivotRightBars) ?? pivotRightBars
        orderBlockUseWicks = try values.decodeIfPresent(Bool.self, forKey: .orderBlockUseWicks) ?? orderBlockUseWicks
        gapAtrMultiplier = try values.decodeIfPresent(Double.self, forKey: .gapAtrMultiplier) ?? gapAtrMultiplier
        requirePriceVoidGaps = try values.decodeIfPresent(Bool.self, forKey: .requirePriceVoidGaps)
            ?? requirePriceVoidGaps
        breakBufferTicks = try values.decodeIfPresent(Int.self, forKey: .breakBufferTicks) ?? breakBufferTicks
        zoneMitigationPercent = try values.decodeIfPresent(Double.self, forKey: .zoneMitigationPercent)
            ?? zoneMitigationPercent
        minimumTick = try values.decodeIfPresent(Double.self, forKey: .minimumTick) ?? minimumTick
        showFlippedOrigins = try values.decodeIfPresent(Bool.self, forKey: .showFlippedOrigins) ?? showFlippedOrigins
        showAllBrokenLevels = try values.decodeIfPresent(Bool.self, forKey: .showAllBrokenLevels)
            ?? showAllBrokenLevels
        hidePooledLines = try values.decodeIfPresent(Bool.self, forKey: .hidePooledLines) ?? hidePooledLines
        poolClusterThreshold = try values.decodeIfPresent(Int.self, forKey: .poolClusterThreshold)
            ?? poolClusterThreshold
        poolAtrFactor = try values.decodeIfPresent(Double.self, forKey: .poolAtrFactor) ?? poolAtrFactor
        maxSupportPools = try values.decodeIfPresent(Int.self, forKey: .maxSupportPools) ?? maxSupportPools
        maxResistancePools = try values.decodeIfPresent(Int.self, forKey: .maxResistancePools) ?? maxResistancePools
        showIfvg = try values.decodeIfPresent(Bool.self, forKey: .showIfvg) ?? showIfvg
        showFvgCe = try values.decodeIfPresent(Bool.self, forKey: .showFvgCe) ?? showFvgCe
        showFvgLabels = try values.decodeIfPresent(Bool.self, forKey: .showFvgLabels) ?? showFvgLabels
        fvgFillMode = try values.decodeIfPresent(String.self, forKey: .fvgFillMode) ?? fvgFillMode
        fvgFillPercent = try values.decodeIfPresent(Double.self, forKey: .fvgFillPercent) ?? fvgFillPercent
        fvgLookback = try values.decodeIfPresent(Int.self, forKey: .fvgLookback) ?? fvgLookback
        fvgBodyPercent = try values.decodeIfPresent(Double.self, forKey: .fvgBodyPercent) ?? fvgBodyPercent
        fvgWickPercent = try values.decodeIfPresent(Double.self, forKey: .fvgWickPercent) ?? fvgWickPercent
        maxVisibleFvgs = try values.decodeIfPresent(Int.self, forKey: .maxVisibleFvgs) ?? maxVisibleFvgs
        fvgMaxBarsActive = try values.decodeIfPresent(Int.self, forKey: .fvgMaxBarsActive) ?? fvgMaxBarsActive
        fvgMinGapAtr = try values.decodeIfPresent(Double.self, forKey: .fvgMinGapAtr) ?? fvgMinGapAtr
        fvgMinBodyAtr = try values.decodeIfPresent(Double.self, forKey: .fvgMinBodyAtr) ?? fvgMinBodyAtr
        fvgBullishColor = try values.decodeIfPresent(String.self, forKey: .fvgBullishColor) ?? fvgBullishColor
        fvgBearishColor = try values.decodeIfPresent(String.self, forKey: .fvgBearishColor) ?? fvgBearishColor
        fvgCeColor = try values.decodeIfPresent(String.self, forKey: .fvgCeColor) ?? fvgCeColor
        ifvgBullishColor = try values.decodeIfPresent(String.self, forKey: .ifvgBullishColor) ?? ifvgBullishColor
        ifvgBearishColor = try values.decodeIfPresent(String.self, forKey: .ifvgBearishColor) ?? ifvgBearishColor
    }

    static let `default` = UsrSettings()

    var isValid: Bool {
        let colors = [fvgBullishColor, fvgBearishColor, fvgCeColor, ifvgBullishColor, ifvgBearishColor]
        return ["chart", "auto", "4h", "1d", "3d", "1w", "2w", "1m", "custom"].contains(analysisTimeframe)
            && ["touch", "close", "ce", "percent"].contains(fvgFillMode)
            && UsrSettingsBounds.proximityPercent.contains(proximityPercent)
            && UsrSettingsBounds.maxSupportLevels.contains(maxSupportLevels)
            && UsrSettingsBounds.maxResistanceLevels.contains(maxResistanceLevels)
            && UsrTimeframe.parse(customTimeframe) != nil
            && UsrSettingsBounds.maxRecentSignalsTotal.contains(maxRecentSignalsTotal)
            && UsrSettingsBounds.volumeLookback.contains(volumeLookback)
            && UsrSettingsBounds.minimumRelativeVolume.contains(minimumRelativeVolume)
            && UsrSettingsBounds.minimumVolumeZScore.contains(minimumVolumeZScore)
            && UsrSettingsBounds.maxSequenceLength.contains(maxSequenceLength)
            && UsrSettingsBounds.displacementBodyPercent.contains(displacementBodyPercent)
            && UsrSettingsBounds.displacementAtrMultiplier.contains(displacementAtrMultiplier)
            && UsrSettingsBounds.structureLookback.contains(structureLookback)
            && UsrSettingsBounds.pivotLeftBars.contains(pivotLeftBars)
            && UsrSettingsBounds.pivotRightBars.contains(pivotRightBars)
            && UsrSettingsBounds.gapAtrMultiplier.contains(gapAtrMultiplier)
            && UsrSettingsBounds.minimumTick.contains(minimumTick)
            && UsrSettingsBounds.breakBufferTicks.contains(breakBufferTicks)
            && UsrSettingsBounds.zoneMitigationPercent.contains(zoneMitigationPercent)
            && UsrSettingsBounds.poolClusterThreshold.contains(poolClusterThreshold)
            && UsrSettingsBounds.poolAtrFactor.contains(poolAtrFactor)
            && UsrSettingsBounds.maxSupportPools.contains(maxSupportPools)
            && UsrSettingsBounds.maxResistancePools.contains(maxResistancePools)
            && UsrSettingsBounds.fvgFillPercent.contains(fvgFillPercent)
            && UsrSettingsBounds.fvgLookback.contains(fvgLookback)
            && UsrSettingsBounds.fvgBodyPercent.contains(fvgBodyPercent)
            && UsrSettingsBounds.fvgWickPercent.contains(fvgWickPercent)
            && UsrSettingsBounds.maxVisibleFvgs.contains(maxVisibleFvgs)
            && UsrSettingsBounds.fvgMaxBarsActive.contains(fvgMaxBarsActive)
            && UsrSettingsBounds.fvgMinGapAtr.contains(fvgMinGapAtr)
            && UsrSettingsBounds.fvgMinBodyAtr.contains(fvgMinBodyAtr)
            && colors.allSatisfy(ScriptColor.isValid)
    }
}

File: apps/desktop/src/features/chart/ultimateSupportResistance/usrSettings.ts
import { isValidScriptColor } from '../scriptOverlayTypes';
import { parseUsrTimeframeValue } from './usrTimeframe';

export type UsrAnalysisTimeframe =
  'chart' | 'auto' | '4h' | '1d' | '3d' | '1w' | '2w' | '1m' | 'custom';

export type UsrFvgFillMode = 'touch' | 'close' | 'ce' | 'percent';

export interface UsrSettings {
  enabled: boolean;
  enableProximityFilter: boolean;
  proximityPercent: number;
  maxSupportLevels: number;
  maxResistanceLevels: number;
  showLiquidityPools: boolean;
  showFvg: boolean;
  analysisTimeframe: UsrAnalysisTimeframe;
  customTimeframe: string;
  showConfluence: boolean;
  enableSrFlip: boolean;
  showBounceSignals: boolean;
  showSweepSignals: boolean;
  signalRequireQualification: boolean;
  requireConfirmationCandleDirection: boolean;
  cancelOpposingSignal: boolean;
  maxRecentSignalsTotal: number;
  volumeLookback: number;
  minimumRelativeVolume: number;
  minimumVolumeZScore: number;
  sessionAwareVolume: boolean;
  maxSequenceLength: number;
  displacementBodyPercent: number;
  displacementAtrMultiplier: number;
  structureLookback: number;
  pivotLeftBars: number;
  pivotRightBars: number;
  orderBlockUseWicks: boolean;
  gapAtrMultiplier: number;
  requirePriceVoidGaps: boolean;
  breakBufferTicks: number;
  zoneMitigationPercent: number;
  minimumTick: number;
  showFlippedOrigins: boolean;
  showAllBrokenLevels: boolean;
  hidePooledLines: boolean;
  poolClusterThreshold: number;
  poolAtrFactor: number;
  maxSupportPools: number;
  maxResistancePools: number;
  showIfvg: boolean;
  showFvgCe: boolean;
  showFvgLabels: boolean;
  fvgFillMode: UsrFvgFillMode;
  fvgFillPercent: number;
  fvgLookback: number;
  fvgBodyPercent: number;
  fvgWickPercent: number;
  maxVisibleFvgs: number;
  fvgMaxBarsActive: number;
  fvgMinGapAtr: number;
  fvgMinBodyAtr: number;
  fvgBullishColor: string;
  fvgBearishColor: string;
  fvgCeColor: string;
  ifvgBullishColor: string;
  ifvgBearishColor: string;
}

export const DEFAULT_USR_SETTINGS: UsrSettings = {
  enabled: false,
  enableProximityFilter: true,
  proximityPercent: 5,
  maxSupportLevels: 125,
  maxResistanceLevels: 125,
  showLiquidityPools: true,
  showFvg: true,
  analysisTimeframe: 'chart',
  customTimeframe: '60',
  showConfluence: false,
  enableSrFlip: true,
  showBounceSignals: false,
  showSweepSignals: false,
  signalRequireQualification: true,
  requireConfirmationCandleDirection: true,
  cancelOpposingSignal: true,
  maxRecentSignalsTotal: 20,
  volumeLookback: 30,
  minimumRelativeVolume: 1.2,
  minimumVolumeZScore: 0.5,
  sessionAwareVolume: true,
  maxSequenceLength: 12,
  displacementBodyPercent: 60,
  displacementAtrMultiplier: 0.6,
  structureLookback: 5,
  pivotLeftBars: 3,
  pivotRightBars: 1,
  orderBlockUseWicks: false,
  gapAtrMultiplier: 0.3,
  requirePriceVoidGaps: true,
  breakBufferTicks: 1,
  zoneMitigationPercent: 0.95,
  // TradingView supplies syminfo.mintick. The app's candle contract does not,
  // so the equivalent value is explicit and user-correctable.
  minimumTick: 0.01,
  showFlippedOrigins: true,
  showAllBrokenLevels: false,
  hidePooledLines: true,
  poolClusterThreshold: 3,
  poolAtrFactor: 2.5,
  maxSupportPools: 30,
  maxResistancePools: 30,
  showIfvg: true,
  showFvgCe: true,
  showFvgLabels: false,
  fvgFillMode: 'ce',
  fvgFillPercent: 50,
  fvgLookback: 10,
  fvgBodyPercent: 0.36,
  fvgWickPercent: 0.5,
  maxVisibleFvgs: 5,
  fvgMaxBarsActive: 200,
  fvgMinGapAtr: 0.05,
  fvgMinBodyAtr: 0.5,
  fvgBullishColor: 'rgba(1, 199, 31, 0.15)',
  fvgBearishColor: 'rgba(216, 0, 0, 0.15)',
  fvgCeColor: 'rgba(255, 235, 59, 0.30)',
  ifvgBullishColor: 'rgba(255, 152, 0, 0.15)',
  ifvgBearishColor: 'rgba(156, 39, 176, 0.15)',
};

export type UsrNumberBounds = { minimum: number; maximum: number; integer?: boolean };

export const USR_NUMBER_BOUNDS: Record<keyof UsrSettings, UsrNumberBounds | undefined> = {
  enabled: undefined,
  enableProximityFilter: undefined,
  proximityPercent: { minimum: 1, maximum: 50 },
  maxSupportLevels: { minimum: 1, maximum: 500, integer: true },
  maxResistanceLevels: { minimum: 1, maximum: 500, integer: true },
  showLiquidityPools: undefined,
  showFvg: undefined,
  analysisTimeframe: undefined,
  customTimeframe: undefined,
  showConfluence: undefined,
  enableSrFlip: undefined,
  showBounceSignals: undefined,
  showSweepSignals: undefined,
  signalRequireQualification: undefined,
  requireConfirmationCandleDirection: undefined,
  cancelOpposingSignal: undefined,
  maxRecentSignalsTotal: { minimum: 5, maximum: 100, integer: true },
  volumeLookback: { minimum: 10, maximum: 200, integer: true },
  minimumRelativeVolume: { minimum: 1, maximum: 5 },
  minimumVolumeZScore: { minimum: 0, maximum: 5 },
  sessionAwareVolume: undefined,
  maxSequenceLength: { minimum: 2, maximum: 50, integer: true },
  displacementBodyPercent: { minimum: 40, maximum: 95 },
  displacementAtrMultiplier: { minimum: 0.2, maximum: 3 },
  structureLookback: { minimum: 2, maximum: 20, integer: true },
  pivotLeftBars: { minimum: 1, maximum: 10, integer: true },
  pivotRightBars: { minimum: 1, maximum: 5, integer: true },
  orderBlockUseWicks: undefined,
  gapAtrMultiplier: { minimum: 0.05, maximum: 3 },
  requirePriceVoidGaps: undefined,
  breakBufferTicks: { minimum: 1, maximum: 20, integer: true },
  zoneMitigationPercent: { minimum: 0.5, maximum: 1 },
  minimumTick: { minimum: 0.000_001, maximum: 100 },
  showFlippedOrigins: undefined,
  showAllBrokenLevels: undefined,
  hidePooledLines: undefined,
  poolClusterThreshold: { minimum: 2, maximum: 10, integer: true },
  poolAtrFactor: { minimum: 1, maximum: 5 },
  maxSupportPools: { minimum: 1, maximum: 60, integer: true },
  maxResistancePools: { minimum: 1, maximum: 60, integer: true },
  showIfvg: undefined,
  showFvgCe: undefined,
  showFvgLabels: undefined,
  fvgFillMode: undefined,
  fvgFillPercent: { minimum: 10, maximum: 100 },
  fvgLookback: { minimum: 3, maximum: 50, integer: true },
  fvgBodyPercent: { minimum: 0.05, maximum: 3 },
  fvgWickPercent: { minimum: 0, maximum: 2 },
  maxVisibleFvgs: { minimum: 1, maximum: 15, integer: true },
  fvgMaxBarsActive: { minimum: 10, maximum: 500, integer: true },
  fvgMinGapAtr: { minimum: 0, maximum: 1 },
  fvgMinBodyAtr: { minimum: 0, maximum: 3 },
  fvgBullishColor: undefined,
  fvgBearishColor: undefined,
  fvgCeColor: undefined,
  ifvgBullishColor: undefined,
  ifvgBearishColor: undefined,
};

const ANALYSIS_TIMEFRAMES = new Set<UsrAnalysisTimeframe>([
  'chart',
  'auto',
  '4h',
  '1d',
  '3d',
  '1w',
  '2w',
  '1m',
  'custom',
]);
const FVG_FILL_MODES = new Set<UsrFvgFillMode>(['touch', 'close', 'ce', 'percent']);

export function validateUsrSettings(candidate: unknown): UsrSettings {
  if (!candidate || typeof candidate !== 'object' || Array.isArray(candidate)) {
    throw new Error('Ultimate S/R settings must be an object.');
  }
  const value = candidate as Record<string, unknown>;
  const expected = Object.keys(DEFAULT_USR_SETTINGS);
  if (Object.keys(value).some((key) => !expected.includes(key))) {
    throw new Error('Ultimate S/R settings contain an unknown field.');
  }
  const merged = { ...DEFAULT_USR_SETTINGS, ...value } as UsrSettings;
  for (const key of expected as Array<keyof UsrSettings>) {
    const defaultValue = DEFAULT_USR_SETTINGS[key];
    const field = merged[key];
    const bounds = USR_NUMBER_BOUNDS[key];
    if (bounds) {
      if (
        typeof field !== 'number' ||
        !Number.isFinite(field) ||
        field < bounds.minimum ||
        field > bounds.maximum ||
        (bounds.integer === true && !Number.isInteger(field))
      ) {
        throw new Error(`Ultimate S/R setting ${key} is invalid.`);
      }
    } else if (typeof defaultValue === 'boolean' && typeof field !== 'boolean') {
      throw new Error(`Ultimate S/R setting ${key} must be boolean.`);
    }
  }
  if (!ANALYSIS_TIMEFRAMES.has(merged.analysisTimeframe)) {
    throw new Error('Ultimate S/R analysis timeframe is invalid.');
  }
  if (
    typeof merged.customTimeframe !== 'string' ||
    !parseUsrTimeframeValue(merged.customTimeframe)
  ) {
    throw new Error('Ultimate S/R custom timeframe is invalid.');
  }
  if (!FVG_FILL_MODES.has(merged.fvgFillMode)) {
    throw new Error('Ultimate S/R FVG fill mode is invalid.');
  }
  for (const color of [
    merged.fvgBullishColor,
    merged.fvgBearishColor,
    merged.fvgCeColor,
    merged.ifvgBullishColor,
    merged.ifvgBearishColor,
  ]) {
    if (typeof color !== 'string' || !isValidScriptColor(color)) {
      throw new Error('Ultimate S/R color is invalid.');
    }
  }
  return merged;
}

export function decodeUsrSettings(candidate: unknown): UsrSettings {
  return validateUsrSettings(candidate);
}
